File: SOR.scala
import scala.io.Source
import javax.print.DocFlavor.INPUT_STREAM
object SOR {
	def main(args: Array[String]) {
        var A = readCSV()
        
        var B = Array.ofDim[Double](A.length,1)
       
          for(i <- 0 to B.length-1){
                  B(i)(0) = A(i)(A(1).length-1)
          }
        var C = Array.ofDim[Double](A.length,A(1).length-1)

        for(i <- 0 to A.length-1){
             for(j <- 0 to A(1).length-2){
                C(i)(j) = A(i)(j)
             }
          }
          for(i <- 0 to C.length-1){
             for(j <- 0 to C(1).length-1){
                print(C(i)(j))
             }
             println()
          }
	          
        var X0:Array[Double] = new Array[Double](A.length)
        
        var matMul = C.zip(X0) map (_.zipped map (_ * _)) map (_.sum)
        
        val bMap = B.toMap

        var subMat = matMul.map{ case (k,v) => (k, v - bMap.getOrElse(k,0)) }.filter(_._2 != 0)
        // var tol = (A,X0).zipped.map((x,y) => (x,B).zipped.map(_*_).sum )

        // print(tol)
        var tolerancia: Double = 0.00001
        var iteracionMax: Int = 100
        var Omega: Double = 0.5
        
        var n: Int = A.size    
        var m: Int = A.size

        var X = X0
        
        var diferencia = Array(1.0, 1.0, 1.0)
        
        var errado = tolerancia*2
        var iteracion : Int = 0
        
        while (!(errado <= tolerancia || iteracion > iteracionMax)){
            for(fila <- 0 to n - 1){
                var suma: Double = 0
                var sigma: Double = 0
                for(columna <- 0 to m - 1){
                    if(fila != columna){
                        sigma = sigma + ((A(fila)(columna)) * (X(columna))) 
                    }
                }
                var phi = (1-Omega) * X(fila) + (Omega / A(fila)(fila)) * (B(fila)(0)-sigma)
                diferencia.update(fila,(phi - (X(fila))).abs)
                X.update(fila, phi)
            }
            errado = diferencia.max
            iteracion = iteracion + 1
        }       
        
        var Xt = Array(Array(X(0)))
        for(i <- 1 to X.length-1){
            Xt ++= Array(Array(X(i)))
        }

        if(iteracion > iteracionMax){
            Xt = Array(Array(0),
                     Array(0),
                     Array(0))
        }

        var revision = Array(0.0, 0.0, 0.0)
        // A*Xt
        for(i <- 0 to A.length - 1){
            for(j <- 0 to Xt(0).length - 1){
                for(k <- 0 to Xt.length - 1){
                    revision.update(i, (revision(i)+(A(i)(k) * Xt(k)(j)))) 
                }
            }
        }  
        
        print("Error: ")
        print(errado)
        print(" --- Iteraciones hechas: ")
        print(iteracion)
        println("")
        println("Respuesta X : ")
        for(i <- 0 to Xt.length-1){
            print(Xt(i)(0) + " \n")
        }
        println("")
        println("Verificar A*X=B: ")
        for(i <- 0 to revision.length-1){
            print(revision(i) + " \n")
        }
        println("")
    }

    def readCSV() : Array[Array[Double]] = {
    io.Source.fromFile("test.csv")
    .getLines()
    .map(_.split(",").map(_.trim.toDouble))
    .toArray
}


}

File: SOR.go
package main

import (
	"fmt"
	"math"
	"time"
	// "github.com/gonum/matrix/mat64"
)

func main() {
	start := time.Now()

	A := [][]float64{
		{3., -0.1, -0.2},
		{0.1, 7, -0.3},
		{0.3, -0.2, 10},
	}

	B := [][]float64{
		{7.85},
		{-19.3},
		{71.4},
	}

	// Ax := mat64.Matrix(A)
	// norm := mat64.Norm(A, 2)
	X0 := [][]float64{
		{0., 0., 0.},
	}

	var tolerancia float64 = 0.0000001
	var iteracionMax int = 100

	var n int = len(A)
	var m int = len(A)
	var omega float64 = 0.5
	var X = X0

	diferencia := [][]float64{
		{1, 1, 1},
	}

	var errado = tolerancia * 2
	var iteracion int = 0

	for !(errado <= tolerancia || iteracion > iteracionMax) {
		for fila := 0; fila < n; fila++ {
			var suma float64 = 0
			for columna := 0; columna < m; columna++ {
				if fila != columna {
					suma = (A[fila][columna]) * (X[0][columna])
				}
			}
			var nuevo = (1.0-omega)*X[0][fila] + (omega/A[fila][fila])*(B[fila][0]-suma)

			diferencia[0][fila] = math.Abs(nuevo - X[0][fila])

			X[0][fila] = nuevo

		}
		var max float64
		for _, j := range diferencia {
			var n float64
			for _, v := range j {
				if v > n {
					n = v
					max = n
				}
			}
		}
		errado = max
		iteracion = iteracion + 1
	}

	if iteracion > iteracionMax {
		X = [][]float64{{0., 0., 0.}}
	}

	//revision, _ := multiply(A, X)

	fmt.Println("Error:", errado, "--- Iteraciones hechas:", iteracion)
	fmt.Println("")
	fmt.Println("Respuesta X: ", X)
	fmt.Println("")
	//fmt.Println("Verificar A*X=B: ", revision)
	fmt.Println("")

	elapsed := time.Since(start)
	fmt.Println("Tiempo de ejecucion:", elapsed)
}

func transpose(x [][]float64) [][]float64 {
	out := make([][]float64, len(x[0]))
	for i := 0; i < len(x); i += 1 {
		for j := 0; j < len(x[0]); j += 1 {
			out[j] = append(out[j], x[i][j])
		}
	}
	return out
}
